include/oneapi/tbb/concurrent_queue.h

```diff
@@ -137,25 +137,29 @@ class concurrent_queue {
     }
 
     concurrent_queue& operator=( const concurrent_queue& other ) {
-        //TODO: implement support for std::allocator_traits::propagate_on_container_copy_assignment
-        if (my_queue_representation != other.my_queue_representation) {
-            clear();
+        if (my_queue_representation == other.my_queue_representation)
+            return *this;
+        clear();
+        if (queue_allocator_traits::propagate_on_container_move_assignment::value) {
             my_allocator = other.my_allocator;
-            my_queue_representation->assign(*other.my_queue_representation, my_allocator, copy_construct_item);
         }
+        my_queue_representation->assign(*other.my_queue_representation, my_allocator, copy_construct_item);
         return *this;
     }
 
     concurrent_queue& operator=( concurrent_queue&& other ) {
-        //TODO: implement support for std::allocator_traits::propagate_on_container_move_assignment
-        if (my_queue_representation != other.my_queue_representation) {
-            clear();
+        if (my_queue_representation == other.my_queue_representation)
+            return *this;
+        clear();
+        if (queue_allocator_traits::propagate_on_container_move_assignment::value) {
+            my_allocator = std::move(other.my_allocator);
+            internal_swap(other);
```

Comment:
I am afraid that with the full support for stateful allocators, propagation and uses-allocator construction, we cannot use `internal_swap` as an implementation for move semantics anymore:

Consider `this->my_allocator` to be stateful allocator. 
After `internal_swap`, the memory allocated by `*this` would be transferred to `other` but the allocator with the correct state would not be transfered to `*this` (and it cannot be since the standard requires us to move-construct allocators. 
Because of these, we will need a "fair" move semantics for both constructor and the assignment operator

Reply:
my_queue_representation is allocated by r1::cache_aligned_allocate instead of my_allocator and all memory allocated by my_allocator has been deallocated at line 153, so I don’t think this is an issue.